File: ts/project/.sdk/tm/rb/feature/rbac_feature.rb
# ProjectName SDK rbac feature
#
# Client-side role/permission enforcement. Before an operation resolves its
# endpoint, the required permission for that entity+operation is checked
# against the permissions the client holds; a disallowed call is
# short-circuited with an "rbac_denied" error and never touches the
# network. Required permissions come from "rules" (keyed by
# "<entity>.<op>", "<op>", or "*"); the default when no rule matches is
# controlled by "deny" (default: allow when unspecified). Held permissions
# are the "permissions" list (a "*" grants everything).

require_relative 'base_feature'

class ProjectNameRbacFeature < ProjectNameBaseFeature
  def initialize
    super
    @version = "0.0.1"
    @name = "rbac"
    # Inactive until init (feature_init only fires init when active).
    @active = false
    @client = nil
    @options = {}
    @granted = {}
  end

  def init(ctx, options)
    @client = ctx.client
    @options = options.is_a?(Hash) ? options : {}
    @active = @options["active"] == true

    @granted = {}
    perms = @options["permissions"]
    if perms.is_a?(Array)
      perms.each { |p| @granted[p] = true }
    end
  end

  def PrePoint(ctx)
    return unless @active

    required = _required(ctx)
    if required.nil?
      # No rule: honour the default policy.
      return _reject(ctx, "<default-deny>") if @options["deny"] == true
      return
    end

    if @granted["*"] || @granted[required]
      _track(ctx, required, true)
      return
    end

    _reject(ctx, required)
  end

  def _required(ctx)
    rules = @options["rules"] || {}
    entity = ""
    if ctx.entity && ctx.entity.respond_to?(:get_name)
      entity = ctx.entity.get_name
    elsif ctx.op
      entity = ctx.op.entity
    end
    opname = ctx.op ? ctx.op.name : ""

    return rules["#{entity}.#{opname}"] unless rules["#{entity}.#{opname}"].nil?
    return rules[opname] unless rules[opname].nil?
    return rules["*"] unless rules["*"].nil?
    nil
  end

  def _reject(ctx, required)
    _track(ctx, required, false)
    opname = (ctx.op && !ctx.op.name.empty?) ? ctx.op.name : "?"
    err = ctx.make_error("rbac_denied",
      "Permission \"#{required}\" required for operation \"#{opname}\"")
    # Short-circuit endpoint resolution; make_point surfaces this error
    # before any network use.
    ctx.out["point"] = err
    err
  end

  def _track(ctx, required, allowed)
    track = @client.instance_variable_get(:@_rbac)
    if track.nil?
      track = { "allowed" => 0, "denied" => 0, "last" => nil }
      @client.instance_variable_set(:@_rbac, track)
    end
    track[allowed ? "allowed" : "denied"] += 1
    track["last"] = {
      "required" => required,
      "allowed" => allowed,
      "op" => ctx.op ? ctx.op.name : nil,
    }
  end
end

File: ts/project/.sdk/tm/py/feature/rbac_feature.py
# ProjectName SDK rbac feature

from __future__ import annotations

from feature.base_feature import ProjectNameBaseFeature


# Client-side role/permission enforcement. Before an operation resolves its
# endpoint, the required permission for that entity+operation is checked
# against the permissions the client holds; a disallowed call is
# short-circuited with an `rbac_denied` error placed in ctx.out["point"]
# (the pipeline's make_point surfaces it) and never touches the network.
# Required permissions come from `rules` (keyed by `<entity>.<op>`, `<op>`,
# or `*`); the default when no rule matches is controlled by `deny`
# (default: allow when unspecified). Held permissions are the `permissions`
# list (a `*` grants everything).
class ProjectNameRbacFeature(ProjectNameBaseFeature):
    def __init__(self):
        super().__init__()
        self.version = "0.0.1"
        self.name = "rbac"
        self.active = True
        self.client = None
        self.options = {}
        self.granted = {}

    def init(self, ctx, options):
        self.client = ctx.client
        self.options = options if isinstance(options, dict) else {}

        if self.options.get("active") is True:
            self.active = True
        else:
            self.active = False

        self.granted = {}
        perms = self.options.get("permissions") or []
        for perm in perms:
            self.granted[perm] = True

    def PrePoint(self, ctx):
        if not self.active:
            return

        required = self._required(ctx)
        if required is None:
            # No rule: honour the default policy.
            if self.options.get("deny") is True:
                return self._reject(ctx, "<default-deny>")
            return

        if self.granted.get("*") is True or self.granted.get(required) is True:
            self._track(ctx, required, True)
            return

        return self._reject(ctx, required)

    def _required(self, ctx):
        rules = self.options.get("rules") or {}

        entity = ""
        ent = ctx.entity
        if ent is not None and hasattr(ent, "get_name") and callable(ent.get_name):
            entity = ent.get_name()
        if entity == "" and ctx.op is not None and isinstance(ctx.op.entity, str):
            entity = ctx.op.entity

        opname = ""
        if ctx.op is not None and isinstance(ctx.op.name, str):
            opname = ctx.op.name

        if rules.get(entity + "." + opname) is not None:
            return rules[entity + "." + opname]
        if rules.get(opname) is not None:
            return rules[opname]
        if rules.get("*") is not None:
            return rules["*"]
        return None

    def _reject(self, ctx, required):
        self._track(ctx, required, False)
        opname = "?"
        if ctx.op is not None and isinstance(ctx.op.name, str) and ctx.op.name != "":
            opname = ctx.op.name
        err = ctx.make_error("rbac_denied",
            'Permission "' + required + '" required for operation "' + opname + '"')
        # Short-circuit endpoint resolution; the pipeline surfaces this error.
        ctx.out["point"] = err
        return err

    def _track(self, ctx, required, allowed):
        client = self.client
        track = getattr(client, "_rbac", None)
        if track is None:
            track = {"allowed": 0, "denied": 0, "last": None}
            client._rbac = track
        track["allowed" if allowed else "denied"] += 1
        track["last"] = {
            "required": required,
            "allowed": allowed,
            "op": ctx.op.name if ctx.op is not None else None,
        }
